src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java

```diff
@@ -605,23 +612,43 @@ default ProgrammingExercise findByIdWithTemplateAndSolutionParticipationTeamAssi
     }
 
     /**
-     * Find a programming exercise by its id, with eagerly loaded template and solution participation, submissions and results
+     * Finds a programming exercise by its id including the submissions for its solution and template participation.
      *
-     * @param programmingExerciseId of the programming exercise.
-     * @return The programming exercise related to the given id
-     * @throws EntityNotFoundException the programming exercise could not be found.
+     * @param exerciseId            The id of the programming exercise.
+     * @param withSubmissionResults True, if the results of the template and solution should be included as well.
+     * @param withGradingCriteria   True, if the grading instructions of the exercise should be included as well.
+     * @return A programming exercise that has the given id.
+     * @throws EntityNotFoundException In case no exercise with the given id exists.
      */
     @NotNull
-    default ProgrammingExercise findByIdWithTemplateAndSolutionParticipationSubmissionsAndResultsAndAuxiliaryRepositoriesElseThrow(long programmingExerciseId)
-            throws EntityNotFoundException {
-        Optional<ProgrammingExercise> programmingExercise = findWithTemplateAndSolutionParticipationSubmissionsAndResultsAndAuxiliaryRepositoriesById(programmingExerciseId);
-        return programmingExercise.orElseThrow(() -> new EntityNotFoundException("Programming Exercise", programmingExerciseId));
+    default ProgrammingExercise findByIdWithTemplateAndSolutionParticipationSubmissionsAndAuxiliaryRepositoriesElseThrow(long exerciseId, boolean withSubmissionResults,
+            boolean withGradingCriteria) throws EntityNotFoundException {
+        final Optional<ProgrammingExercise> programmingExercise;
+
+        if (withGradingCriteria) {
+            if (withSubmissionResults) {
+                programmingExercise = findWithTemplateAndSolutionParticipationSubmissionsAndResultsAndAuxiliaryRepositoriesAndGradingCriteriaById(exerciseId);
+            }
+            else {
+                programmingExercise = findWithTemplateAndSolutionParticipationAndAuxiliaryRepositoriesAndGradingCriteriaById(exerciseId);
+            }
+        }
+        else {
+            if (withSubmissionResults) {
+                programmingExercise = findWithTemplateAndSolutionParticipationSubmissionsAndResultsAndAuxiliaryRepositoriesById(exerciseId);
+            }
+            else {
+                programmingExercise = findWithTemplateAndSolutionParticipationAndAuxiliaryRepositoriesById(exerciseId);
+            }
+        }
```

Comment:
Optional / maybe followup:

This is now a bit nested and we have 4 different JPA-Queries that are (only?) used for this purpose. A CiteriaBuilder for this purpose would be a great refactoring. See https://docs.artemis.cit.tum.de/dev/guidelines/criteria-builder.html and / or #6043.